client/__tests__/e2e/e2e.test.ts

```diff
@@ -85,448 +78,477 @@ const setToRemoveFrom = "empty_this_geneset";
     const histBrushableAreaId = `histogram-${data.clip.metadata}-plot-brushable-area`;
     const coords = await calcDragCoordinates(
       histBrushableAreaId,
-      data.clip["coordinates-as-percent"]
+      data.clip["coordinates-as-percent"],
+      page
     );
-    await drag(histBrushableAreaId, coords.start, coords.end);
-    const cellCount = await getCellSetCount(1);
+    await drag(histBrushableAreaId, coords.start, coords.end, page);
+    const cellCount = await getCellSetCount(1, page);
     expect(cellCount).toBe(data.clip.count);
 
     // ensure categorical data appears properly
-    for (const label of Object.keys(data.categorical)) {
-      const element = await getOneElementInnerHTML(
-        getTestId(`category-${label}`)
-      );
+    for (const label of Object.keys(
+      data.categorical
+    ) as (keyof typeof data.categorical)[]) {
+      const element = await page.getByTestId(`category-${label}`).innerHTML();
 
       expect(element).toMatchSnapshot();
 
-      await clickOn(`${label}:category-expand`);
+      await page.getByTestId(`${label}:category-expand`).click();
 
-      const categories = await getAllCategoriesAndCounts(label);
+      const categories = await getAllCategoriesAndCounts(label, page);
 
       expect(Object.keys(categories)).toMatchObject(
-        // @ts-expect-error ts-migrate(7053) FIXME: Element implicitly has an 'any' type because expre... Remove this comment to see the full error message
         Object.keys(data.categorical[label])
       );
 
       expect(Object.values(categories)).toMatchObject(
-        // @ts-expect-error ts-migrate(7053) FIXME: Element implicitly has an 'any' type because expre... Remove this comment to see the full error message
         Object.values(data.categorical[label])
       );
-    }    
+    }
   });
 });
 
 // interact with UI elements just that they do not break
 describe("ui elements don't error", () => {
-  test("color by", async () => {
-    await goToPage(pageUrl);
-
+  test("color by", async ({ page }) => {
+    await goToPage(page);
     const allLabels = [
       ...Object.keys(data.categorical),
       ...Object.keys(data.continuous),
     ];
 
     for (const label of allLabels) {
-      await clickOn(`colorby-${label}`);
+      await page.getByTestId(`colorby-${label}`).click();
     }
   });
 
-  test("pan and zoom", async () => {
-    await goToPage(pageUrl);
-
-    await clickOn("mode-pan-zoom");
+  test("pan and zoom", async ({ page }) => {
+    await goToPage(page);
+    await page.getByTestId("mode-pan-zoom").click();
     const panCoords = await calcDragCoordinates(
       "layout-graph",
-      data.pan["coordinates-as-percent"]
+      data.pan["coordinates-as-percent"],
+      page
     );
 
-    await drag("layout-graph", panCoords.start, panCoords.end, false);
+    await drag("layout-graph", panCoords.start, panCoords.end, page);
 
     await page.evaluate("window.scrollBy(0, 1000);");
```

Comment:
same for this one - not sure how this test can detect when pan/zoom breaks, since it just performs actions but not actually asserting the UI actually changes somehow

Maybe a ticket to actually assert it if possible!

If we can't use element attributes to detect changes, we also now have early access to Chromatics e2e test for visual testing 😄 !

If it'll make these types of testing easier, we should create a ticket to set it up!

https://www.chromatic.com/blog/e2e-visual-tests-sneak-peek/

Reply:
#716